synapse/storage/data_stores/main/events.py

```diff
@@ -1957,6 +1973,105 @@ def insert_labels_for_event_txn(
+            txn (LoggingTransaction): The database transaction to use.
+            event_id (str): The event ID the expiry timestamp is associated with.
+            expiry_ts (int): The timestamp at which to expire (delete) the event.
+        """
+        return self._simple_insert_txn(
+            txn=txn,
+            table="event_expiry",
+            values={"event_id": event_id, "expiry_ts": expiry_ts},
+        )
+
+    def expire_event(self, event):
+        """Redact an event that has expired, and delete its associated expiry timestamp.
+
+        Args:
+             event (events.EventBase): The event to delete.
+        """
+        # Prune the event's dict then convert it to JSON.
+        pruned_json = encode_json(prune_event_dict(event.get_dict()))
+
+        def delete_expired_event_txn(txn):
+            # Update the event_json table to replace the event's JSON with the pruned
+            # JSON.
+            self._censor_event_txn(txn, event.event_id, pruned_json)
+
+            # Delete the expiry timestamp associated with this event from the database.
+            self._simple_delete_txn(
+                txn, table="event_expiry", keyvalues={"event_id": event.event_id}
+            )
+
+            # We need to invalidate the event cache entry for this event because we
+            # changed its content in the database.
+            txn.call_after(self._get_event_cache.invalidate, (event.event_id,))
+            # Send that invalidation to replication so that other workers also invalidate
+            # the event cache.
+            self._send_invalidation_to_replication(
+                txn, "_get_event_cache", (event.event_id,)
+            )
+
+        return self.runInteraction("delete_expired_event", delete_expired_event_txn)
+
+    def delete_event_expiry(self, event_id):
+        """Delete the expiry timestamp associated with an event ID without deleting the
+        actual event.
+
+        Args:
+            event_id (str): The event ID to delete the associated expiry timestamp of.
+        """
+        return self._simple_delete(
+            table="event_expiry",
+            keyvalues={"event_id": event_id},
+            desc="delete_event_expiry",
+        )
+
+    def get_events_to_expire(self):
+        """Retrieve the IDs of the events we have an expiry timestamp for, along with
+        said timestamp.
+
+        Returns:
+            A list of dicts, each containing an event_id and an expiry_ts.
+        """
+        return self._simple_select_list(
+            table="event_expiry",
+            keyvalues=None,
+            retcols=["event_id", "expiry_ts"],
+            desc="get_events_to_expire",
+        )
+
+    def get_next_event_to_expire(self):
+        """Retrieve the entry with the lowest expiry timestamp in the event_expiry
+        table, or None if there's no more event to expire.
+
+        Returns:
+            A dict with an event_id and an expiry_ts of there's at least one row in the
```

Comment:
```suggestion
            A dict with an event_id and an expiry_ts if there's at least one row in the
```

It'd be easier and cleaner to return a 2-sequence imho:

```
txn.execute(...)
return txn.fetchone()
```